### PR TITLE
Support external plugins

### PR DESCRIPTION
**Pull request by hkaiser**
A couple of minor changes allowing for external libraries to load their own plugins through HPX's plugin infrastructure. 

These are changes required by Phylanx.

### CODE REVIEW

File: hpx/plugins/plugin_registry.hpp [hunk 1]
```diff
@@ -57,11 +57,11 @@ namespace hpx { namespace plugins
...
         {
             using namespace boost::assign;
-            fillini += std::string("[hpx.plugins.") +
+            fillini += std::string("[" HPX_PLUGIN_CONFIG_SECTION ".plugins.") +
                 unique_plugin_name<plugin_registry>::call() + "]";
             fillini += "name = " HPX_PLUGIN_STRING;
-            fillini += std::string("path = ") +
-                util::find_prefixes("/hpx", HPX_PLUGIN_STRING);
+            fillini += std::string("path = ") + util::find_prefixes(
+                    "/" HPX_PLUGIN_PATH_SUFFIX, HPX_PLUGIN_STRING);
```
Comment (sithhell) [inline hpx/plugins/plugin_registry.hpp]: Wouldn't it be better to have `HPX_PLUGIN_CONFIG_SECTION` and `HPX_PLUGIN_PATH_SUFFIX` as a parameter to the plugin registry instead? Making it relying on the global `config.hpp` header looks error prone. The following should work as well: https://wandbox.org/permlink/4emy0DIwSOzxmSjs

And provides a much cleaner interface for the module declarations.

Comment (hkaiser) [reply hpx/plugins/plugin_registry.hpp]: This will break NVCC most likely, but we might be able to keep this code out of view of this compiler.

Comment (hkaiser) [reply hpx/plugins/plugin_registry.hpp]: Done.

File: hpx/plugins/plugin_registry.hpp [hunk 2]
```diff
@@ -89,15 +87,32 @@ namespace hpx { namespace plugins
...
+    typedef hpx::plugins::plugin_registry<PluginType, __##pluginname##_string, \
+        __##pluginname##_section, __##pluginname##_suffix>                     \
+        __##pluginname##_plugin_registry_type;                                 \
+    HPX_REGISTER_PLUGIN_BASE_REGISTRY(                                         \
+        __##pluginname##_plugin_registry_type, pluginname)                     \
+    HPX_DEF_UNIQUE_PLUGIN_NAME(                                                \
+        __##pluginname##_plugin_registry_type, pluginname)                     \
+    template struct hpx::plugins::plugin_registry<PluginType,                  \
+        __##pluginname##_string, __##pluginname##_section,                     \
+        __##pluginname##_suffix>;                                              \
```
Comment (sithhell) [inline hpx/plugins/plugin_registry.hpp]: Should we rather use `HPX_PP_CAT` to concatenate the PP tokens?

Comment (hkaiser) [reply hpx/plugins/plugin_registry.hpp]: This is only necessary if the parameters `pluginname` is a macro itself.. Not sure if it's worth the trouble.

**Comment by sithhell** [comment]
http://cdash.cscs.ch/testDetails.php?test=913265&build=6076

I think this test failure is related.

**Comment by hkaiser** [comment]
> http://cdash.cscs.ch/testDetails.php?test=913265&build=6076
> I think this test failure is related.

Yes, this is related but was fixed by the latest commit.

File: hpx/plugins/plugin_registry.hpp [hunk 3]
```diff
@@ -89,15 +87,32 @@ namespace hpx { namespace plugins
...
-    template struct hpx::plugins::plugin_registry<PluginType >;               \
-/**/
+#define HPX_REGISTER_PLUGIN_REGISTRY_2(PluginType, pluginname)                 \
+    HPX_REGISTER_PLUGIN_REGISTRY_5(                                            \
+        PluginType, pluginname, HPX_PLUGIN_NAME, "hpx", "hpx")                 \
+    /**/
+#define HPX_REGISTER_PLUGIN_REGISTRY_4(                                        \
+    PluginType, pluginname, pluginsection, pluginsuffix)                       \
+    HPX_REGISTER_PLUGIN_REGISTRY_5(PluginType, pluginname, HPX_PLUGIN_NAME,    \
+        pluginsection, "pluginsuffix")                                         \
```
Comment (sithhell) [inline hpx/plugins/plugin_registry.hpp]: why is `pluginsuffix` quoted here?

File: hpx/plugins/plugin_registry.hpp (hunk 2 above)
Comment (sithhell) [reply hpx/plugins/plugin_registry.hpp]: Ok, we'll see how it goes.

File: hpx/plugins/plugin_registry.hpp (hunk 3 above)
Comment (hkaiser) [reply hpx/plugins/plugin_registry.hpp]: Yep, caught that yesterday as well.... Should be fine now.

**Comment by hkaiser** [comment]
@sithhell canm we go ahead with this now?

**Pull request merged**